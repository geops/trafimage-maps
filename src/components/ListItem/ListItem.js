import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  /**
   * Css class of the li tag.
   * @ignore
   */
  className: PropTypes.string,

  /**
   * Item of the list.
   * @ignore
   */
  item: PropTypes.any.isRequired,

  /**
   * Children content of the button.
   * @ignore
   */
  children: PropTypes.node.isRequired,

  /**
   * Function triggered when you select the item.
   * @ignore
   */
  onSelect: PropTypes.func,

  /**
   * Function triggered when you press a key.
   * @ignore
   */
  onKeyDown: PropTypes.func,
};

const defaultProps = {
  className: 'tm-list-item',
  onSelect: () => {},
  onKeyDown: () => {},
};

/**
 * This component displays a `<li>` tag.
 * @ignore
 */
class ListItem extends PureComponent {
  render() {
    const { className, children, item, onSelect, onKeyDown } = this.props;
    return (
      <li
        className={className}
        role="menuitem"
        tabIndex="0"
        onClick={(e) => {
          onSelect(e, item);
        }}
        onKeyPress={(e) => e.which === 13 && onSelect(e, item)}
        onKeyDown={(e) => {
          onKeyDown(e);
        }}
      >
        {children}
      </li>
    );
  }
}

ListItem.propTypes = propTypes;
ListItem.defaultProps = defaultProps;

export default ListItem;
