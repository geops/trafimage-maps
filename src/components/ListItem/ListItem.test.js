import React from "react";
import { render, fireEvent } from "@testing-library/react";
import ListItem from "./ListItem";

describe("ListItem", () => {
  const item = {
    label: "foo",
  };
  const children = <p>bar</p>;

  describe("when no properties are set", () => {
    let spy = null;

    beforeEach(() => {
      window.console.error = jest.fn().mockImplementation(() => {});
      spy = jest.spyOn(window.console, "error");
    });

    afterEach(() => {
      spy.mockRestore();
      window.console.error.mockRestore();
    });

    test("displays 1 error for required property ", () => {
      render(<ListItem />);
      expect(spy).toHaveBeenCalledTimes(2);
    });

    test("matches snapshot", () => {
      const { container } = render(<ListItem />);
      expect(container.innerHTML).toMatchSnapshot();
    });

    test("calls default onSelect and onKeyDown without errors.", () => {
      const { container } = render(<ListItem item={item}>{children}</ListItem>);
      fireEvent.click(container.querySelector("p"));
      fireEvent.keyDown(container.querySelector("p"));
      expect(spy).toHaveBeenCalledTimes(0);
    });
  });

  describe("when properties are set", () => {
    test("matches snapshot", () => {
      const { container } = render(
        <ListItem item={item} onSelect={() => {}}>
          {children}
        </ListItem>,
      );
      expect(container.innerHTML).toMatchSnapshot();
    });

    test("calls onSelect property on click.", () => {
      const fn = jest.fn();
      const { container } = render(
        <ListItem item={item} onSelect={fn}>
          {children}
        </ListItem>,
      );
      fireEvent.click(container.querySelector("p"));
      expect(fn).toHaveBeenCalledTimes(1);
    });

    test("calls onSelect property on click.", () => {
      const fn = jest.fn();
      const { container } = render(
        <ListItem item={item} onSelect={fn}>
          {children}
        </ListItem>,
      );
      fireEvent.click(container.querySelector("p"));
      expect(fn).toHaveBeenCalledTimes(1);
    });

    test("calls onSelect property on enter key press.", () => {
      const fn = jest.fn();
      const { container } = render(
        <ListItem item={item} onSelect={fn}>
          {children}
        </ListItem>,
      );
      fireEvent.keyPress(container.querySelector("p"), {
        which: 13,
        keyCode: 13,
      });
      expect(fn).toHaveBeenCalledTimes(1);
    });

    test("shouldn't calls onSelect property on key press other than enter", () => {
      const fn = jest.fn();
      const { container } = render(
        <ListItem item={item} onSelect={fn}>
          {children}
        </ListItem>,
      );
      fireEvent.keyPress(container.querySelector("p"), {
        which: 14,
        keyCode: 14,
      });
      expect(fn).toHaveBeenCalledTimes(0);
    });

    test("calls onKeyDown property on key down", () => {
      const fn = jest.fn();
      const { container } = render(
        <ListItem item={item} onKeyDown={fn}>
          {children}
        </ListItem>,
      );
      fireEvent.keyDown(container.querySelector("p"));
      expect(fn).toHaveBeenCalledTimes(1);
    });
  });
});
