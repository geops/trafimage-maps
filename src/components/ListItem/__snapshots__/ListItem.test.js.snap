// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem when no properties are set matches snapshot 1`] = `
<li class="tm-list-item"
    role="menuitem"
    tabindex="0"
>
</li>
`;

exports[`ListItem when properties are set matches snapshot 1`] = `
<li class="tm-list-item"
    role="menuitem"
    tabindex="0"
>
  <p>
    bar
  </p>
</li>
`;
